core/coreapi/pubsub.go

```diff
@@ -0,0 +1,161 @@
+package coreapi
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	core "github.com/ipfs/go-ipfs/core"
+	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
+	caopts "github.com/ipfs/go-ipfs/core/coreapi/interface/options"
+
+	cid "gx/ipfs/QmPSQnBKM9g7BaUcZCvswUJVscQ1ipjmwxN5PXCjkp9EQ7/go-cid"
+	floodsub "gx/ipfs/QmPbYVFhKxamZAN9MyrQMDeoGYa6zkQkhAPguwFfzxPM1J/go-libp2p-floodsub"
+	peer "gx/ipfs/QmbNepETomvmXfz1X5pHNFD2QuPqnqi47dTd94QJWSorQ3/go-libp2p-peer"
+	pstore "gx/ipfs/QmfAQMFpgDU2U4BXG64qVr8HSiictfWvkSBz7Y2oDj65st/go-libp2p-peerstore"
+)
+
+type PubSubAPI CoreAPI
+
+type pubSubSubscription struct {
+	subscription *floodsub.Subscription
+}
+
+type pubSubMessage struct {
+	msg *floodsub.Message
+}
+
+func (api *PubSubAPI) Ls(ctx context.Context) ([]string, error) {
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	return api.node.Floodsub.GetTopics(), nil
+}
+
+func (api *PubSubAPI) Peers(ctx context.Context, opts ...caopts.PubSubPeersOption) ([]peer.ID, error) {
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	settings, err := caopts.PubSubPeersOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	peers := api.node.Floodsub.ListPeers(settings.Topic)
+	out := make([]peer.ID, len(peers))
+
+	for i, peer := range peers {
+		out[i] = peer
+	}
+
+	return out, nil
+}
+
+func (api *PubSubAPI) Publish(ctx context.Context, topic string, data []byte) error {
+	if err := api.checkNode(); err != nil {
+		return err
+	}
+
+	return api.node.Floodsub.Publish(topic, data)
+}
+
+func (api *PubSubAPI) Subscribe(ctx context.Context, topic string, opts ...caopts.PubSubSubscribeOption) (coreiface.PubSubSubscription, error) {
+	options, err := caopts.PubSubSubscribeOptions(opts...)
+
+	if err := api.checkNode(); err != nil {
+		return nil, err
+	}
+
+	sub, err := api.node.Floodsub.Subscribe(topic)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.Discover {
+		go func() {
+			blk, err := api.core().Block().Put(ctx, strings.NewReader("floodsub:"+topic))
+			if err != nil {
+				log.Error("pubsub discovery: ", err)
+				return
+			}
+
+			connectToPubSubPeers(ctx, api.node, blk.Path().Cid())
+		}()
+	}
+
+	return &pubSubSubscription{sub}, nil
+}
+
+func connectToPubSubPeers(ctx context.Context, n *core.IpfsNode, cid cid.Cid) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	provs := n.Routing.FindProvidersAsync(ctx, cid, 10)
+	wg := &sync.WaitGroup{}
```

Comment:
Nit: We don't actually need to allocate here. We can just use `var wg sync.WaitGroup`.

(not really important but still)